crates/core-subsystem/core-resolver/src/context/provider/jwt/authenticator.rs

```diff
@@ -0,0 +1,114 @@
+use std::cell::RefCell;
+use std::env;
+
+use jsonwebtoken::errors::ErrorKind;
+use jsonwebtoken::{decode, DecodingKey, TokenData, Validation};
+use serde_json::Value;
+use tracing::warn;
+
+use thiserror::Error;
+
+use crate::context::error::ContextExtractionError;
+use crate::context::request::Request;
+
+use super::jwks::{JwksEndpoint, JwksExtractionError};
+
+pub(super) const EXO_JWT_SECRET: &str = "EXO_JWT_SECRET";
+pub(super) const EXO_JWKS_ENDPOINT: &str = "EXO_JWKS_ENDPOINT";
+
+// we spawn many resolvers concurrently in integration tests
+thread_local! {
+    pub static LOCAL_JWT_SECRET: RefCell<Option<String>> = RefCell::new(None);
+    pub static LOCAL_JWKS_URL: RefCell<Option<String>> = RefCell::new(None);
+}
+
+/// Authenticator with information about how to validate JWT tokens
+/// It can be either a secret or a JWKS endpoint
+pub enum JwtAuthenticator {
+    Secret(String),
+    Endpoint(JwksEndpoint),
+}
+
+#[derive(Debug, Error)]
+pub(super) enum JwtAuthenticationError {
+    #[error("Token extraction error `{0}`")]
+    JsonWebToken(#[from] jsonwebtoken::errors::Error),
+
+    #[error("JWKS extraction error `{0}`")]
+    JwksExtractionError(#[from] JwksExtractionError),
+}
+
+impl JwtAuthenticator {
+    pub fn new_from_env() -> Option<Self> {
+        let secret = LOCAL_JWT_SECRET.with(|local_jwt_secret| {
+            local_jwt_secret
+                .borrow()
+                .clone()
+                .or_else(|| env::var(EXO_JWT_SECRET).ok())
+        });
+
+        let jwks_url = LOCAL_JWKS_URL.with(|url| {
+            url.borrow()
+                .clone()
+                .or_else(|| env::var(EXO_JWKS_ENDPOINT).ok())
+        });
+
+        match (secret, jwks_url) {
+            (Some(secret), None) => Some(JwtAuthenticator::Secret(secret)),
+            (None, Some(jwks_url)) => Some(JwtAuthenticator::Endpoint(JwksEndpoint::new(jwks_url))),
+            (Some(_), Some(_)) => {
+                warn!("Both {EXO_JWT_SECRET} and {EXO_JWKS_ENDPOINT} are set. JWT authentication will not be enabled.");
+                None
```

Comment:
Exiting here might be preferable since the server is clearly misconfigured.

Reply:
Good point. Changed it to make it a configuration error.